File: expressjs/src/engine/workflow-handler.ts
import { Router, type Express, type Request, type Response, type NextFunction, type RequestHandler } from "express";
import type { Store } from "../store/postgres.js";
import type { Registry } from "../metadata/registry.js";
import { AppError } from "./errors.js";
import {
  resolveWorkflowAction,
  loadWorkflowInstance,
  listPendingInstances,
} from "./workflow.js";
import { getInstrumenter } from "../instrument/instrument.js";

type AsyncHandler = (req: Request, res: Response, next: NextFunction) => Promise<void>;

function asyncHandler(fn: AsyncHandler) {
  return (req: Request, res: Response, next: NextFunction) => {
    fn(req, res, next).catch(next);
  };
}

export class WorkflowHandler {
  private store: Store;
  private registry: Registry;

  constructor(store: Store, registry: Registry) {
    this.store = store;
    this.registry = registry;
  }

  getPendingInstances = asyncHandler(async (_req: Request, res: Response) => {
    const instances = await listPendingInstances(this.store);
    res.json({ data: instances });
  });

  getInstance = asyncHandler(async (req: Request, res: Response) => {
    const id = req.params.id;
    try {
      const instance = await loadWorkflowInstance(this.store, id);
      res.json({ data: instance });
    } catch {
      throw new AppError("NOT_FOUND", 404, `Workflow instance not found: ${id}`);
    }
  });

  approveInstance = asyncHandler(async (req: Request, res: Response) => {
    const id = req.params.id;
    const userID = (req.headers["x-user-id"] as string) || "system";
    const span = getInstrumenter().startSpan("workflow", "handler", "workflow.approve");
    span.setMetadata("instance_id", id);
    span.setMetadata("user_id", userID);
    try {
      const instance = await resolveWorkflowAction(
        this.store,
        this.registry,
        id,
        "approved",
        userID,
      );
      span.setStatus("ok");
      res.json({ data: instance });
    } catch (err: any) {
      span.setStatus("error");
      span.setMetadata("error", err.message);
      if (err.message?.includes("not found")) {
        throw new AppError("NOT_FOUND", 404, err.message);
      }
      if (err.message?.includes("not running") || err.message?.includes("not an approval")) {
        throw new AppError("INVALID_STATE", 422, err.message);
      }
      throw err;
    } finally {
      span.end();
    }
  });

  rejectInstance = asyncHandler(async (req: Request, res: Response) => {
    const id = req.params.id;
    const userID = (req.headers["x-user-id"] as string) || "system";
    const span = getInstrumenter().startSpan("workflow", "handler", "workflow.reject");
    span.setMetadata("instance_id", id);
    span.setMetadata("user_id", userID);
    try {
      const instance = await resolveWorkflowAction(
        this.store,
        this.registry,
        id,
        "rejected",
        userID,
      );
      span.setStatus("ok");
      res.json({ data: instance });
    } catch (err: any) {
      span.setStatus("error");
      span.setMetadata("error", err.message);
      if (err.message?.includes("not found")) {
        throw new AppError("NOT_FOUND", 404, err.message);
      }
      if (err.message?.includes("not running") || err.message?.includes("not an approval")) {
        throw new AppError("INVALID_STATE", 422, err.message);
      }
      throw err;
    } finally {
      span.end();
    }
  });
}

export function registerWorkflowRoutes(
  app: Express,
  handler: WorkflowHandler,
  ...middleware: RequestHandler[]
): void {
  const router = Router();

  router.get("/pending", handler.getPendingInstances);
  router.get("/:id", handler.getInstance);
  router.post("/:id/approve", handler.approveInstance);
  router.post("/:id/reject", handler.rejectInstance);

  app.use("/api/_workflows", ...middleware, router);
}

File: backend/golang/internal/engine/workflow_handler.go
package engine

import (
	"github.com/gofiber/fiber/v2"

	"rocket-backend/internal/instrument"
	"rocket-backend/internal/metadata"
	"rocket-backend/internal/store"
)

// WorkflowHandler handles workflow runtime HTTP endpoints.
type WorkflowHandler struct {
	store    *store.Store
	registry *metadata.Registry
}

func NewWorkflowHandler(s *store.Store, reg *metadata.Registry) *WorkflowHandler {
	return &WorkflowHandler{store: s, registry: reg}
}

// RegisterWorkflowRoutes adds workflow runtime routes.
// Must be registered AFTER admin routes but BEFORE dynamic entity routes.
func RegisterWorkflowRoutes(app *fiber.App, h *WorkflowHandler, middleware ...fiber.Handler) {
	wf := app.Group("/api/_workflows", middleware...)
	wf.Get("/pending", h.ListPending)
	wf.Get("/:id", h.GetInstance)
	wf.Post("/:id/approve", h.Approve)
	wf.Post("/:id/reject", h.Reject)
	wf.Delete("/:id", h.Delete)
}

func (h *WorkflowHandler) GetInstance(c *fiber.Ctx) error {
	id := c.Params("id")
	instance, err := loadWorkflowInstance(c.Context(), h.store, id)
	if err != nil {
		return NewAppError("NOT_FOUND", 404, "Workflow instance not found: "+id)
	}
	return c.JSON(fiber.Map{"data": instance})
}

func (h *WorkflowHandler) ListPending(c *fiber.Ctx) error {
	instances, err := ListPendingInstances(c.Context(), h.store)
	if err != nil {
		return NewAppError("INTERNAL_ERROR", 500, "Failed to list pending instances")
	}
	if instances == nil {
		instances = []*metadata.WorkflowInstance{}
	}
	return c.JSON(fiber.Map{"data": instances})
}

func (h *WorkflowHandler) Approve(c *fiber.Ctx) error {
	ctx := c.UserContext()
	ctx, span := instrument.GetInstrumenter(ctx).StartSpan(ctx, "workflow", "handler", "workflow.approve")
	defer span.End()
	c.SetUserContext(ctx)

	id := c.Params("id")
	span.SetMetadata("instance_id", id)
	userID := c.Get("X-User-ID", "anonymous") // Until Auth is implemented

	instance, err := ResolveWorkflowAction(c.Context(), h.store, h.registry, id, "approved", userID)
	if err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return NewAppError("VALIDATION_FAILED", 422, err.Error())
	}

	span.SetStatus("ok")
	return c.JSON(fiber.Map{"data": instance})
}

func (h *WorkflowHandler) Delete(c *fiber.Ctx) error {
	id := c.Params("id")
	if err := DeleteWorkflowInstance(c.Context(), h.store, id); err != nil {
		return NewAppError("NOT_FOUND", 404, "Workflow instance not found: "+id)
	}
	return c.JSON(fiber.Map{"data": fiber.Map{"deleted": true}})
}

func (h *WorkflowHandler) Reject(c *fiber.Ctx) error {
	ctx := c.UserContext()
	ctx, span := instrument.GetInstrumenter(ctx).StartSpan(ctx, "workflow", "handler", "workflow.reject")
	defer span.End()
	c.SetUserContext(ctx)

	id := c.Params("id")
	span.SetMetadata("instance_id", id)
	userID := c.Get("X-User-ID", "anonymous")

	instance, err := ResolveWorkflowAction(c.Context(), h.store, h.registry, id, "rejected", userID)
	if err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return NewAppError("VALIDATION_FAILED", 422, err.Error())
	}

	span.SetStatus("ok")
	return c.JSON(fiber.Map{"data": instance})
}
